Print binary values of floats

thk123 opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add test for printing different float values in the trace

Commit:
Adapt test to check for binary in trace

Commit:
Add additional tests for more types

Commit:
Pull out function for producing the full_lhs_value

Commit:
Adapt full_lhs_value to output binary for floating values

This is a precise representation of the floats, which can be useful

Commit:
Adapt the XSD to include the binary output

Also updat the LaTeX and markdown documentation to include the new
attribute

Commit:
Add folder readme for doc/assets

Includes dependencies for builing the xml_spec LaTeX, as well as running
the trace through the specification.

Commit:
Referenced readme in cprover docs

The links were not relative, so updated to be relative. Mentioned that
you can find more in the readme.

thk123 commented on doc/assets/xml_spec.md:
```diff
@@ -192,7 +200,15 @@ always present
+        <xs:complexType>
+          <xs:simpleContent>
+            <xs:extension base="xs:string">
+              <xs:attribute name="binary" type="xs:string"/>
```
:pick: this could be made more precise using a XSD type where we limit this to be only [01]+

thk123 reviewed (commented):
Just leaving it here, happy to do it if others think it is of any use

peterschrammel commented on src/goto-programs/xml_goto_trace.cpp:
```diff
@@ -19,10 +19,38 @@ Author: Daniel Kroening
+  if(full_lhs_value_includes_binary(step, ns))
+  {
+    const auto width = to_floatbv_type(step.full_lhs_value.type()).get_width();
+    const auto &binary_representation = integer2binary(
```
```suggestion
    const auto binary_representation = integer2binary(
```

peterschrammel commented on src/goto-programs/xml_goto_trace.cpp:
```diff
@@ -19,10 +19,38 @@ Author: Daniel Kroening
+  xmlt full_lhs_value{"full_lhs_value"};
+
+  auto lhs_object = step.get_lhs_object();
+  irep_idt identifier =
```
```suggestion
  const irep_idt identifier =
```

peterschrammel commented on src/goto-programs/xml_goto_trace.cpp:
```diff
@@ -19,10 +19,38 @@ Author: Daniel Kroening
+{
+  xmlt full_lhs_value{"full_lhs_value"};
+
+  auto lhs_object = step.get_lhs_object();
```
```suggestion
  const auto &lhs_object = step.get_lhs_object();
```

peterschrammel commented on regression/cbmc/float-trace/test.desc:
```diff
@@ -0,0 +1,32 @@
+<full_lhs_value binary="[01]{64}">\+NAN</full_lhs_value>
+--
+--
+Add tests to check the binary value is printed for floating
```
```suggestion
Checks that the binary value is printed for floating point types
```

peterschrammel commented on doc/cprover-manual/cbmc-tutorial.md:
```diff
@@ -138,9 +138,11 @@ possible output format.
-and can be build by `pdflatex -shell-escape xml_spec.tex`. Alternatively, you
-view it in Markdown [here](../../../assets/xml_spec).
+Specification](../assets/xml_spec.tex)
+and can be build by `pdflatex -shell-escape xml_spec.tex`. See the [README
```
```suggestion
and can be built by `pdflatex -shell-escape xml_spec.tex`. See the [README
```

peterschrammel commented on doc/cprover-manual/cbmc-tutorial.md:
```diff
@@ -138,9 +138,11 @@ possible output format.
+and can be build by `pdflatex -shell-escape xml_spec.tex`. See the [README
+](../assets/README.md#xml_spec.tex) for details.
+
+Alternatively, you view it in [Markdown](../assets/xml_spec.md).
```
```suggestion
Alternatively, you can view it in [Markdown](../assets/xml_spec.md).
```

peterschrammel commented on doc/assets/xml_spec.tex:
```diff
@@ -147,9 +147,15 @@ \subsection{Trace Steps in XML}
-\item \texttt{full\_lhs\_value}: a constant with the new value
+\item \texttt{full\_lhs\_value}: a constant with the new value.
+  \begin{itemize}
+    \item if the type is a floating point, there will be an attribute
```
```suggestion
    \item if the type is a floating point number type, there will be an attribute
```

peterschrammel commented on doc/assets/xml_spec.md:
```diff
@@ -168,7 +168,9 @@ always present
 
--   `full_lhs_value`: a constant with the new value
+-   `full_lhs_value`: a constant with the new value.
+     - if the type is a floating point, there will be an attribute`binary` 
```
```suggestion
     - if the type is a floating point number type, there will be an attribute`binary` 
```

peterschrammel commented on doc/assets/README.md:
```diff
@@ -0,0 +1,28 @@
+
+# XML Spec
+
+These are files relating to the specification the trace when using the XML.
```
```suggestion
The folllowing files relate to the specification of the XML trace output when using the `--xml-ui` flag.
```

peterschrammel commented on doc/assets/README.md:
```diff
@@ -0,0 +1,28 @@
+
+## xml_spec.tex
+
+This is a LaTeX document for the specification. It can be compiled to pdf using:
```
```suggestion
This is a LaTeX document describing the specification of the trace. It can be compiled to pdf using:
```

Commit:
Tidy up full_lhs_value

Applied modern coding standards to pulled out code

Co-authored-by: Peter Schrammel <peterschrammel@users.noreply.github.com>

thk123 commented:
@peterschrammel comments addressed and squashed

NlightNFotis commented on doc/cprover-manual/cbmc-tutorial.md:
```diff
@@ -348,4 +350,3 @@ comes with a small set of definitions, which includes functions such as
 
 We also have a [list of interesting applications of
 CBMC](http://www.cprover.org/cbmc/applications/).
-
```
Shouldn't there be a newline here?

thk123 replied:
This was auto-removed on save, because there are two new lines at the end of the file

NlightNFotis reviewed (approved):
LGTM

thk123 commented:
Failure is real - on Windows the complex variable name is prefixed with a _ , I guess this is to do with the name of the variable being reserved on windows so I'll just change the name as not part of the test

Commit:
Rename variables in test to avoid clash on windows

complex is renamed in the trace to _complex on Windows. Therefore adapt
the test to use a non-reserved name since not important to the test.

thk123 commented:
clang format was an apt get failure - restared
